### ISSUE TITLE
Validation is too slow

### ISSUE THREAD

**Issue by athas** [gradbench/gradbench#141]
Gradbench supports validation, but currently all the validation is for the eval to run the PyTorch implementation and compare the result. This has the following downsides:

1. It is inherently inefficient because when running *N* tools with some eval, you also end up running the PyTorch implementation *N* times.
2. The PyTorch implementations are somewhat slow, which makes (1) a significant obstacle for larger workloads.

The basic idea of executing a golden implementation on demand has the problem that it essentially has to run in the same Docker image as the eval. This makes it annoying to simply swap out the PyTorch implementation with some other one, such as the Manual ones from ADBench, as these come with lots of C++ baggage.

This could be addressed by extending the protocol  to allow the eval to communicate with *two* tools (one of which is the "golden" one), but that would be rather complex, so let's not do that. Instead, it should be possible to pass an eval a JSON file that is the result of a previous execution, and contains the golden results, and use that for validation. That JSON file can be computed using whatever tool we trust (hopefully a fast one). I think that was brought up in the initial discussion of the design for validation as well.

**Comment by samestep** [comment]
Personally I'm not a huge fan of the added complexity of having the eval communicate with multiple tools, and I'd be more inclined to just go the route of using the Manual implementations from ADBench inside of the eval itself. I know you mentioned that it's a bit annoying to do this, but it's not too bad, right? Especially since we're going to include Manual under `tools` anyway, per #136.

**Comment by athas** [comment]
We might not have Manual for future evals. In fact, we almost certainly will not. AD is mostly useful for those problems that are completely intractable to differentiate by hand.

**Comment by samestep** [comment]
Yep that is fair. Still though, in general can't we just always use some fast tool for validation in any eval? That means the slowdown is at most 2x. Or we could even just save the expected outputs when we build the Docker image itself, if we really want to save time. All these seem better to me than adding extra complexity to the protocol itself.

**Comment by athas** [comment]
I would prefer to pass in the expected outputs as a command line argument to the eval (which would then have to be copied into the Docker image, but I think that is not so difficult). The specifics can be worked out. I definitely agree that this should be done without modifying the protocol.

**Comment by athas** [comment]
I still think passing in reference results from a previous execution is the best way to do this. but instead of passing it to the eval directly, how about passing it to the `gradbench` program, which could then feed it to the eval as a part of an initial message? This would require a protocol extension, but it is simpler than fiddling with Docker file mapping (and also works transparently when running evals outside of Docker). The eval can do whatever it wants with the "reference log" (including nothing), so it doesn't force us to do anything in particular.

One downside of this approach is that putting *all* results into a single line of JSON will result in a very large line of JSON (easily multiple GiB for some evals), and I don't know how well Python's JSON parser will cope with it. An alternative is to still feed a reference log to `gradbench`, and whenever `gradbench` is about to transmit a tool response to a request from the eval, it will check whether the reference log also contains a response to that request, and if so, add it as a field to the tool response. This also does not require any Docker-specific code, but it does require more logic in `gradbench` itself.

Both of these options maintain the property of keeping the validation logic itself in the eval.

**Comment by samestep** [comment]
Ah I see; when I woke up this morning I was thinking that we could just punt to the eval and do something like this:

```
./gradbench eval foo -- --golden golden.jsonl
```

But that wouldn't quite work because of Docker as you mentioned, so instead we'd need to add filesystem args to the `eval` subcommand itself, resulting in something like this:

```
./gradbench eval foo --file golden.jsonl -- --golden golden.jsonl
```

This still isn't _horrible_ though. I think I'd probably still prefer it to modifying the protocol?

**Comment by athas** [comment]
I don't have any particular preference. If the intermediary can transparently deal with the Docker issues so users don't have to know or care, then that is likely the best solution. I'm not averse to protocol changes, but I'm always a bit wary of transferring enormous values through JSON.

**Comment by athas** [comment]
Closed as inherent to the preferred design (although a minor improvement would be to use a faster tool than pytorch).

**Comment by samestep** [comment]
To clarify (as far as I understand), summarizing the conversation in #196: the current decision is that we plan to keep the status quo architecture for validation, but may do one or more of the following:

1. Choose a faster reference tool.
2. Add flags like `--no-validate` or `--golden` to evals (not the central CLI), the latter of which wouldn't work in Docker without some additional configuration.

We could also precompute reference values when building Docker images, but the downside is that that solution might make it more annoying to run the evals outside Docker. Or, we could commit precomputed values to the repo, but the downside is that those files might be large.

**Comment by athas** [comment]
I took a very quick shot at it, but it's not entirely straightforward, as you need to compile the C++ code in advance - you can't just call the actual function from the wrapper module (e.g. `ht.py`); you need to call `compile` first. Not impossible, but somewhat tedious, and ties implementation details of the golden module into the eval. For example, you should only call `compile` *once*, as it recompiles from scratch on every invocation, which may take a rather long time.

**Comment by samestep** [comment]
Closing this now that #264 is merged.
